
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta name="google-site-verification" content="ptuXe7ex0jPvowLOm_BRWnxrV-EfJeXDTA7zF0aLuzk" />
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

<!-- Begin Jekyll SEO tag v2.8.0 -->
<title>RNAseqFast | A R package for quick RNAseq DE analysis</title>
<meta name="generator" content="Jekyll v3.9.3" />
<meta property="og:title" content="RNAseqFast" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A R package for quick RNAseq DE analysis" />
<meta property="og:description" content="A R package for quick RNAseq DE analysis" />
<link rel="canonical" href="https://fengjiang94.github.io/RNAseqFast/" />
<meta property="og:url" content="https://fengjiang94.github.io/RNAseqFast/" />
<meta property="og:site_name" content="RNAseqFast" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="RNAseqFast" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebSite","description":"A R package for quick RNAseq DE analysis","headline":"RNAseqFast","name":"RNAseqFast","url":"https://fengjiang94.github.io/RNAseqFast/"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="stylesheet" href="/RNAseqFast/assets/css/style.css?v=90f38f390b3ef7383533542322709ad43effcb61">
    <!-- start custom head snippets, customize with your own _includes/head-custom.html file -->

<!-- Setup Google Analytics -->



<!-- You can set your favicon here -->
<!-- link rel="shortcut icon" type="image/x-icon" href="/RNAseqFast/favicon.ico" -->

<!-- end custom head snippets -->

  </head>
  <body>
    <div class="container-lg px-3 my-5 markdown-body">
      

      
<h1 id="rnaseqfast">RNAseqFast</h1>
<p>An R package containing functions for basic RNAseq differential expression analysis and GO analysis.</p>

<p>This package was orignally designed for the downstream analysis after running <a href="https://github.com/FengJiang94/RNAseqPE">RNAseqPE</a>, yet can be used for any mouse or human RNA-seq data.</p>

<p>To use with RNA-seq data generated using other pipelines, please make sure the input files follow the same format as the example inputs.</p>

<p>Example inputs:
<em>FeatureCounts</em> :Standard outputs of <a href="https://subread.sourceforge.net/featureCounts.html">featureCounts</a>.
<em>colData.csv</em>: Simple information.
<em>counts_matrix.csv</em>: A mnerged matrix counting the raw gene counts.</p>

<p>This package was designed for data analysis in the following paper, and please consider citing it:</p>

<ol>
  <li>Jiang, F., Hedaya, O.M., Khor, E., Wu, J., Auguste, M., &amp; Yao, P. (2023). <a href="https://academic.oup.com/nar/article/51/11/5831/7147500">RNA binding protein PRRC2B mediates translation of specific mRNAs and regulates cell cycle progression</a>. <em>Nucleic Acids Research</em>, 51(11), 5831â€“5846</li>
</ol>

<p>Some example figures generated by this package:</p>

<p><img src="/RNAseqFast/image/Picture1.png" alt="" /></p>

<h2 id="software-prerequisites">Software prerequisites</h2>

<p>The following R (&gt;= 3.5.0) packages are required:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>DESeq2
EnhancedVolcano
clusterProfiler
ggplot2
org.Hs.eg.db
org.Mm.eg.db
pheatmap
RColorBrewer
</code></pre></div></div>

<p>To perform GO analysis, <em>org.Mm.eg.db</em>(mouse) and <em>org.Hs.eg.db</em>(human) are required.</p>

<h2 id="installation">Installation</h2>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(devtools)
install_github("FengJiang94/RNAseqFast")

library(RNAseqFast)
</code></pre></div></div>

<h2 id="quick-start">Quick start</h2>

<p>This package contains the following four functions:</p>

<ul>
  <li>
    <p><strong>MergeFeatureCounts</strong>: A function to merge <a href="https://subread.sourceforge.net/featureCounts.html">featureCounts</a> outputs into one table.</p>
  </li>
  <li>
    <p><strong>CPM_c</strong> and <strong>TPM_c</strong>: Two functions to calculate CPM and TPM. <a href="https://www.reneshbedre.com/blog/expression_units.html">Definition of CPM and TPM</a>.</p>
  </li>
  <li>
    <p><strong>RNAseqDEGs</strong>: The core function to perform differential expression analysis from raw counts.
This function outputs:
A heatmap of sample correlation
A heatmap of sample distance
A volcano plot showing differential expressed genes
A MA plot
A heatmap showing the top 20 significant differential expressed genes
A matrix containing log2FoldChange and Pvalue for each gene 
A matrix containing log2FoldChange and Pvalue for each coding gene (if lnc=T) 
A matrix containing log2FoldChange and Pvalue for each non coding gene (if lnc=T)</p>
  </li>
  <li>
    <p><strong>GOnKegg</strong>: A function to perform GO and Kegg analysis given a gene list.</p>
  </li>
</ul>

<p>Here is an example  to perform DE analysis for RNA-seq:
Input data used here can be downloaded from <a href="https://github.com/FengJiang94/ExampleData">Example inputs</a>.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(RNAseqFast)

# Set your working directory
# All results will be generated here
CurrPath &lt;- "Example outputs"
setwd(CurrPath)

# Set path to the input data
## Path to the fold containing featureCounts outputs
FeatureCounts &lt;- "Example inputs/FeatureCounts"
## Path to the colData
colDataPath &lt;- "Example inputs/colData.csv"

# Load colData
colData &lt;- read.csv(colDataPath, header = T)

# Merge FeatureCounts outputs in to one table
counts_matrix &lt;- MergeFeatureCounts(FeatureCounts, 
                                    Assay ="Example outputs/RNAseq", 
                                    colDataPath = colDataPath, 
                                    write_to_file = T)

# Calculate CPM
CPM_c(counts_matrix = counts_matrix,
      Assay = "Example outputs/RNAseq")

# Calculate TPM
TPM_c(Assay = "Example outputs/RNAseq",
      counts_matrix = counts_matrix,
      lengthdata = "hg38",
      Plots = T,
      coldata = colData)

# DE analysis 
## Here we compare gene expressin between "apple" and "banana" using total RNA extracted from cell
## This is the most common analysis for RNA-seq
## We subset the colData to include only samples we need for the comparsion. 
## This can aslo be done using the select1 argument 
colData_total &lt;- colData[colData$condition_2 == "total",]
## We convert colData to fators
colData_total$condition &lt;- factor(colData_total$condition, levels = c("banana", "apple"))
## levels = c("Banana", "Apple") make sure the comparsion is made as Apple/Banana
setwd("Example outputs")
ApplevsBanana &lt;- RNAseqDEGs(coldata = colData_total,
                            counts_matrix = counts_matrix,
                            filename = "RNAseq",
                            By = "padj", 
                            lfc = 1, 
                            LRT = F,
                            Design = ~ condition,
                            intgroup = "condition",
                            lnc = T,
                            shrinkcoef = 2,
                            lnc_ref = "human",
                            plot_coding = T,
                            tpm_file = "RNAseq_tpm.csv")
## check result                            
ApplevsBanana[[3]]

## Another example for DE analysis
## polysome-seq analysis
## Here we compare the transaltion efficiency (TE) between "apple" and "banana" using total RNA extracted and polysome associated RNAs from cell.
## This is an old fashined way to calculate TE, yet still used by some labs.
## To do this, colData should have a column telling whether a sample is from total RNA or polysome associated RNA.
colData$condition_2
## This column should be fators.
colData$condition_2 &lt;- factor(colData$condition_2, levels = c("total", "polysome"))
## it is important to set levels = c("total", "polysome") since transaltion efficiency (TE) is calculated by polysome/total
## levels = c("Banana", "Apple") make sure the comparsion is made as Apple/Banana
colData$condition &lt;- factor(colData$condition, levels = c("banana", "apple"))
## analysis
ApplevsBananaTE &lt;- RNAseqDEGs(coldata = colData,
                              counts_matrix = counts_matrix,
                              filename = "TE",
                              By = "padj", 
                              lfc = 1, 
                              LRT = T,
                              Design = ~ condition + condition_2 + condition:condition_2,
                              reduced = ~ condition + condition_2,
                              intgroup = "condition",
                              lnc = T,
                              shrinkcoef = 4,
                              lnc_ref = "human",
                              plot_coding = T,
                              tpm_file = "RNAseq_tpm.csv")

## check results
ApplevsBananaTE[[3]]

# GO and KEGG analysis
## Here we perform GO and KEGG analysis for genes upregulated in Apple by at least 2 folds
ApplevsBanana &lt;- as.data.frame(ApplevsBanana[[3]])
genes &lt;- rownames(ApplevsBanana[ApplevsBanana$log2FoldChange &gt; 1 &amp; ApplevsBanana$padj &lt; 0.05, ])
## Go and KEGG
GOnKEGG(sign_de_gene = genes, filename = "RNAseq")
</code></pre></div></div>
<p>Output for the above script can be downloaded from the <a href="https://github.com/FengJiang94/ExampleData">Example Outputs</a>.</p>



      
    </div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/4.1.0/anchor.min.js" integrity="sha256-lZaRhKri35AyJSypXXs4o6OPFTbTmUoltBbDCbdzegg=" crossorigin="anonymous"></script>
    <script>anchors.add();</script>
  </body>
</html>
