#' Run DEG analysis using raw counts
#'
#' This function takes raw count table and run the full DESeq2 analysis to identify DEGs.
#'
#' This function create a folder containing all the output files.
#'
#' @param colData Sample information text file. This file contain at least three columns:
#' sample_name, short_name, condition_1, condition_2, ... sample_name should be the same as the "stat/sample,csv"
#' in RNAseqPE outputs. Only used if Plots = T
#' @param select1 A vector of short_names of samples to be included in the analysis. By defualt, all samples in the
#' coldata will be included.
#' @param counts_matrix A Merged featureCounts table generated by [MergeFeatureCounts()]. *colnames(counts_matrix)*
#' should be the same as *coldata$short_name*.
#' @param tpm_file Path to the tpm file generated by [TPM_c()].
#' @param filename Prefix of the outputs files.
#' @param LRT True of False, whether to use "LRT" test. If False, Wald test will be used.
#' @param reduced The reduced formula for LRT test. Only used when LRT = T.
#' @param Design The formula or matrix passed to DESeq2. The formula expresses how the counts for each gene depend
#' on the variables in colData File prefix of the outputs".
#' @param shrinkcoef The name or number of the coefficient (LFC) to shrink. See [lfcShrink()]
#' @param shrinktype Type of Shrink to perform, See [lfcShrink()]
#' @param intgroup The column in coldata for coloring samples in the PCA plots
#' @param lnc True of False, whether to separate non coding RNAs versus coding RNAs.
#' @param lnc_ref Refernce for non coding RNAs. It can be either "human", "mouse", or a user-provided
#' vector containing gene names of ncRNAs.
#' @param By "padj" or "pvalue" to use.
#' @param lfc Threshold for |log2FC| when ploting.
#' @param plot_coding True of False, whether to only include coding RNAs in the volcano plot.
#' @param plimit Limitation of Pvalue/Padj to show in the plots. Pvalue < plimit will be convert to plimit.
#' @param xlimit Limitation of |log2FC| to show in the plots. |log2FC| > xlimit will be convert to xlimit.
#' @param fitType Either "parametric", "local", "mean", or "glmGamPoi" for the type of fitting of dispersions
#' to the mean intensity. see [DESeq()] for details.
#' @return A list of three objects: a sample information matrix, dds, and *result(dds)*.
#' @import RColorBrewer
#' @import pheatmap
#' @import EnhancedVolcano
#' @import DESeq2
#' @import ggplot2
#' @export

##################-------DEGs Calculation------######################
RNAseqDEGs<-function(coldata,
                     select1 = c(),
                     counts_matrix,
                     tpm_file = paste(Assay,"_tpm.csv", sep="" ),
                     filename,
                     LRT = FALSE,
                     reduced ="",
                     fitType = "parametric",
                     Design = ~ condition,
                     shrinkcoef = 2,
                     shrinktype ="apeglm",
                     intgroup = "condition",
                     lnc = FALSE,
                     lnc_ref = "human",
                     By = "padj",
                     lfc = 1,
                     plot_coding = FALSE,
                     plimit = 1e-10,
                     xlimit = 10)
{
  # create the output folder
  path <- paste(filename, By, "0.05", lfc, sep = "_")
  dir.create(path = path)
  filename <- paste(path, filename, sep = "/")
  print(paste("store output files in ", path, sep = ""))
  # select samples, By default all samples are used
  if (length(select1) > 0) {
    coldata<-coldata[coldata$short_name %in% select1,]
    print("analysis only includes selected samples")
  }
  #filter the counts_matrix to keep the cols cotaining selected samples
  counts_matrix_selected<-counts_matrix[,coldata$short_name]
  ncol(counts_matrix_selected)
  #filter out low counts genes
  keep<-rowSums(counts_matrix_selected)>=1
  counts_matrix_selected<-counts_matrix_selected[keep,]
  print("DESeq analysis begin")
  #create DEseq object
  dds<-DESeqDataSetFromMatrix(countData = counts_matrix_selected,
                              colData = coldata,
                              design = Design)
  #analysis
  if (LRT == T){
    print("use likelihood ratio test, make sure reduced formula is provided")
    print(paste("run ", fitType, "_test", sep = ""))
    dds<-DESeq(dds, parallel=TRUE, test = 'LRT', reduced = reduced, fitType = fitType)
  }
  else{
    print("use Wald significance tests")
    dds<-DESeq(dds,parallel=TRUE)
  }
  # get the result
  res<-results(dds)
  #set NA p.value, padjust to 1
  res[is.na(res$padj),"padj"]<-1
  res[is.na(res$pvalue),"pvalue"]<-1
  # ordered the results by Padj
  resOrdered <- res[order(res$padj),]
  # save the file
  write.csv(resOrdered,file = paste(filename,".csv",sep=""))
  print("DESeq analysis finished")
  # separate coding versus long non coding RNAs
  #################### long_non_coding_RNA reference #####################
  if (lnc == T)
  {
    print("Separte coding versus noncoding RNAs")
    if (lnc_ref == "human"){
      print("use gencode.v38.long_noncoding_RNAs.gtf as reference")
      lnc_id <- built_in_lnc_ref_hg38
    }
    else if (lnc_ref == "mouse"){
      print("use ensemble m39 lncRNA as reference")
      lnc_id <- built_in_lnc_ref_m39
    }
    else{
      print("use user-provided lncRNA reference")
      lnc_id <- lnc_ref
    }
    # separate coding vs long non coding RNAs
    lnc <- resOrdered[(rownames(resOrdered) %in% lnc_id),]
    cod <- resOrdered[!(rownames(resOrdered) %in% lnc_id),]
    # write the files
    write.csv(lnc, file = paste(filename,"_lnc.csv",sep=""))
    write.csv(cod, file = paste(filename,"_coding.csv",sep=""))
  }

  #normolized_counts
  nor_counts<-counts(dds, normalized = TRUE)
  write.csv(nor_counts,file = paste(filename,"_nor_counts.csv",sep=""))

  #heatmap of sample-sample distence
  print("calculate sample-sample distance, VST used")
  vsd <- vst(dds, blind=FALSE)
  sampleDists <- dist(t(assay(vsd)))
  #distance matrix
  sampleDistMatrix <- as.matrix(sampleDists)
  head(sampleDistMatrix)
  head(sampleDistMatrix)
  #colors
  colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
  #plot and save
  pdf(file = paste(filename,"_dis.pdf",sep=""), width = 12, height = 8)
  p1<-pheatmap(sampleDistMatrix,
               clustering_distance_rows=sampleDists,
               clustering_distance_cols=sampleDists,
               col=colors)
  print(p1)
  dev.off()

  #heatmap of sample-sample correlation
  print("calculate sample-sample correlation, VST used")
  sampleCor <- cor(assay(vsd))
  #correlation matrix
  sampleCorMatrix <- as.matrix(sampleCor)
  #colors
  colors <- colorRampPalette(brewer.pal(9, "Blues") )(255)
  #plot and save
  pdf(file = paste(filename,"_cor.pdf",sep=""), width = 12, height = 8)
  p2 <-pheatmap(sampleCorMatrix, col=colors)
  print(p2)
  dev.off()

  #pca plot for sample-sample distance
  print("generating PCA plots, VST used")
  # color the dots by intgroup
  pca_data <- plotPCA(vsd, intgroup = intgroup, returnData = T, ntop = 5000)
  percentVar <- round(100 * attr(pca_data, "percentVar"))
  pdf(file = paste(filename,"_pca.pdf",sep=""), width = 12, height = 8)
  p <- qplot(pca_data$PC1, pca_data$PC2, color = as.factor(pca_data$group), label = pca_data$name) +
    geom_text(hjust= 0.4, vjust=-1, alpha=0.8, size= 3)
  p<-p + geom_point(size=3) +
    xlab(paste0("PC1: ", percentVar[1], "% variance")) +
    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
    theme(legend.title = element_text(size =1, colour = "transparent"))
  print(p)
  dev.off()

  #MA-plot showing threshold
  print("generating MA-plot plots, lfcshrink used to shrink log2 fold changes")
  #for RNA-seq, shrinkcoef is 2
  #for Translation efficiency analysis, shrinkcoef is 4
  resApeT <- lfcShrink(dds, coef= shrinkcoef, type = shrinktype, lfcThreshold=0)
  # use pvaule or padj depends on By
  if (By == "pvalue"){
    resApeT$padj <- resApeT$pvalue
    print("use raw P value instead of P.adjust")
  }
  pdf(file =paste(filename,"_MA_plot.pdf",sep=""), width = 8, height = 8)
  plotMA(resApeT, ylim=c(-5,5),alpha= 0.05)
  abline(h=c(0-lfc,lfc), col="dodgerblue", lwd=2)
  dev.off()

  # heatmap for top 20 DEGs based on padj or pvalue
  # use TPM to plot
  print("plot a heatmap for top 20 significant DEGs, use TPM")
  if (plot_coding == T)
  {
    print("plot coding genes only")
    res <- cod
  }
  tpm<-read.csv(tpm_file, header=T, row.names = 1)
  DEG<-res[(res$log2FoldChange > lfc |res$log2FoldChange < (-lfc)),]
  if (By == 'padj'){
    print("use raw P value instead of P.adjust")
    resOrder<-order(DEG$padj, decreasing = FALSE)[1:20]
  }
  else{
    resOrder<-order(DEG$pvalue, decreasing = FALSE)[1:20]
  }

  mat<-tpm[rownames(DEG[resOrder,]),coldata$short_name]
  # scale = "row" (mat-rowMeans(mat))/sd(mat)
  # color columns by intgroup
  df <- as.data.frame(coldata[,intgroup])
  rownames(df)<-colnames(mat)
  colnames(df)<-"condition"
  # plot
  pdf(file =paste(filename,"_heatmap_top_20.pdf",sep=""), width = 12, height = 8)
  p<-pheatmap(mat, border_color = NA, cluster_cols = TRUE, scale = "row", annotation_col = df,
              show_rownames=TRUE, cutree_rows = 2, angle_col = 0)
  print(p)
  dev.off()

  #volcano plot
  print("plot a volcano plot, use TPM")
  #set ranges for pvalue and log2FC
  if (nrow(res[res$padj< plimit,])>0)
    res[res$padj< plimit,]$padj<- plimit
  if (nrow(res[res$pvalue<plimit,])>0)
    res[res$pvalue< plimit,]$pvalue<- plimit
  if (nrow(res[res$log2FoldChange > xlimit,])>0)
    res[res$log2FoldChange > xlimit,]$log2FoldChange<- xlimit
  if (nrow(res[res$log2FoldChange< 0-xlimit,])>0)
    res[res$log2FoldChange< 0-xlimit,]$log2FoldChange<- 0-xlimit
  # use P value or Padj depends on "By"
  if (By == "pvalue"){
    keyvals.colour <- ifelse(
      res$log2FoldChange > lfc & res$pvalue < 0.05 , 'red2',
      ifelse(res$log2FoldChange < -lfc & res$pvalue < 0.05, 'red2',
             'grey30'))
             names(keyvals.colour)[keyvals.colour == 'grey30'] <- 'NS'
             names(keyvals.colour)[keyvals.colour == 'red2'] <- 'Significant'
  }
  if (By == "padj"){
    keyvals.colour <- ifelse(
      res$log2FoldChange > lfc & res$padj < 0.05 , 'red2',
      ifelse(res$log2FoldChange < -lfc & res$padj < 0.05, 'red2',
             'grey30'))
             names(keyvals.colour)[keyvals.colour == 'grey30'] <- 'NS'
             names(keyvals.colour)[keyvals.colour == 'red2'] <- 'Significant'
  }
  # plot
  pdf(file =paste(filename,"_volcano.pdf",sep=""),width = 12, height = 8)
  p<-EnhancedVolcano(res,
                     lab = NA,
                     x = 'log2FoldChange',
                     y = By,
                     xlim = c(0-xlimit, xlimit),
                     ylim = c(0, -log10(plimit)),
                     #shapeCustom = keyvals.shape,
                     colCustom = keyvals.colour,
                     title = filename,
                     pCutoff = 0.05,
                     FCcutoff = lfc,
                     #set the x,y labels
                     xlab = bquote(~Log[2]~ 'fold change'),
                     ylab = bquote(~-Log[10]~italic(.(By))),
                     #size of the dots
                     pointSize = 0.5,
                     #color transparency
                     colAlpha = 1,
                     cutoffLineType = 'twodash',
                     cutoffLineWidth = 0.8,
                     legendPosition = "right",
                     drawConnectors = FALSE,
                     legendLabSize = 8,
                     legendIconSize = 2.0,)

  print(p)
  dev.off()
  print(paste("volcano plot only shows genes with |log2FC| < ", xlimit, sep = ""))
  print(paste("volcano plot only shows genes with ",  By,  "> ", plimit, sep = ""))
  print(paste("change xlimit and plimit to include more or less genes"))
  #return two martris, one containg the samples to compare, one containg the results
  return (list(coldata, dds, resOrdered))
}
