#' Convert FeatureCounts to TPM (transcript per million nucleotide)
#'
#' This function converts a Merged featureCounts table into TPM table. Merged featureCounts table is generated by
#'  [MergeFeatureCounts()].
#'
#' This function create two files: a table containing CPM in each sample
#' and a table containing average CPM among all samples. If Plots == T, it also plot a correlation heatmap and a
#' PCA plot based on the TPM
#'
#' @param counts_matrix A Merged featureCounts table generated by [MergeFeatureCounts()].
#' @param Assay File prefix of the outputs".
#' @param lengthdata Data.frame containing the gene length. It could be either "hg38", "mm10", or any customized
#' data.frame. Customized data.frame should have gene names as rownames and only one column containg gene length.
#' @param Plots True of False, whether to plot sample correlations. If Ture, heatmap and PCA plots will be generated
#' with the same file prefix.
#' @param coldata Sample information text file. This file contain at least three columns:
#' sample_name, short_name, condition_1, condition_2, ... sample_name should be the same as the "stat/sample,csv"
#' in RNAseqPE outputs. Only used if Plots = T
#' @param pca_colors Sample groups used to color the PCA points. This should be a vector of factors same length as
#' [ncol(counts_matrix)]. Default is the [ColData$conditions]. Only used if Plots = T.
#' @import RColorBrewer
#' @import pheatmap
#' @import ggplot2
#' @export
#'
###################### TPM Calculation ###########
TPM_c <- function(Assay = Assay,
                  counts_matrix = counts_matrix,
                  lengthdata = "hg38",
                  Plots = T,
                  coldata = coldata,
                  pca_colors = paste(coldata$condition))
{
  # load the gene length
  if (lengthdata == "hg38"){
   lengthdata <- built_in_length_hg38[rownames(counts_matrix),]
   print("use built-in gene length hg38")
  }
  else if (lengthdata == "mm10"){
    lengthdata <- built_in_length_mm10[rownames(counts_matrix),]
    print("use built-in gene length mm10")
  }
  else{
    print("use customized gene length")
    lengthdata <- lengthdata[rownames(counts_matrix),]
  }
  number <- length(lengthdata[is.na(lengthdata)])
  if (number > 0){
    print(paste("Warning: Length information missed for ", number, " out of ", nrow(counts_matrix), " genes", sep=""))
    print("Double check your lengthdata")
  }
  countdata<-counts_matrix
  countdata$Length <- lengthdata
  countdata <- countdata[!is.na(countdata$Length),]
  # calculate TPM
  kb <- countdata$Length / 1000
  rpk <- countdata / kb
  rpk<-rpk[,1:ncol(rpk)-1]
  tpm <- t(t(rpk)/colSums(rpk) * 1000000)
  # calculate average TPM
  avg_tpm <- data.frame(avg_tpm=rowMeans(tpm))
  # save files
  write.csv(tpm, paste(Assay, "_tpm.csv", sep = ""))
  write.csv(avg_tpm, paste(Assay, "_avg_tpm.csv", sep = ""))
  print("calculate TPM done")

  if (Plots == T){
    print("calculate sample-sample correlation heatmap")
    #################heatmap of sample-sample correlation#####################
    tpm <- tpm[rowSums(tpm) >1,]
    sampleCor <- cor(tpm)
    #correlation matrix
    sampleCorMatrix <- as.matrix(sampleCor)
    #colors
    colors <- colorRampPalette(brewer.pal(9, "Blues") )(255)
    #plot and save
    filename<-Assay
    pdf(file = paste(filename,"_all_samples_cor_TPM.pdf",sep=""), width = 12, height = 8)
    p2 <-pheatmap(sampleCorMatrix,col=colors)
    print(p2)
    dev.off()
    print("calculate sample-sample correlation PCA plot")
    #####################pca plot for sample-sample distance######################
    pca_res<-prcomp(t(tpm), center = TRUE,scale. = T)
    percentage <- round(pca_res$sdev^2 / sum(pca_res$sdev^2) * 100, 2)
    #plot and save
    pdf(file = paste(filename,"_all_sample_pca.pdf",sep=""), width = 12, height = 8)
    p <- qplot(x = pca_res$x[,1],
               y = pca_res$x[,2],
               color = pca_colors,
               label = rownames(pca_res$x)) + geom_text(hjust= 0.6, vjust=-1, alpha=0.8, size= 1.7)
    p<-p + geom_point(size=3) +
      xlab(paste0("PC1: ", percentage[1], "% variance")) +
      ylab(paste0("PC2: ", percentage[2], "% variance")) +
      theme(legend.title = element_text(size =1, colour = "transparent"))
    print(p)
    dev.off()
  }
}



